pkg/api_gateway/api_gateway.go

```diff
@@ -0,0 +1,202 @@
+package apigateway
+
+import (
+	"encoding/hex"
+	"errors"
+	"sync"
+
+	"github.com/centrifuge/go-substrate-rpc-client/v4/types"
+	"github.com/rs/zerolog/log"
+	substrate "github.com/threefoldtech/tfchain/clients/tfchain-client-go"
+	"github.com/threefoldtech/zos/pkg"
+)
+
+type apiGateway struct {
+	sub      *substrate.Substrate
+	mu       sync.Mutex
+	identity substrate.Identity
+}
+
+func NewAPIGateway(manager substrate.Manager, identity substrate.Identity) (pkg.APIGateway, error) {
+	sub, err := manager.Substrate()
+	if err != nil {
+		return nil, err
+	}
+	return &apiGateway{sub: sub, mu: sync.Mutex{}, identity: identity}, nil
+}
+
+func (g *apiGateway) CreateNode(node substrate.Node) (uint32, error) {
+	log.Debug().
+		Str("method", "CreateNode").
+		Uint32("twin id", uint32(node.TwinID)).
+		Uint32("farm id", uint32(node.FarmID)).
+		Msg("method called")
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.sub.CreateNode(g.identity, node)
+}
+
+func (g *apiGateway) CreateTwin(relay string, pk []byte) (uint32, error) {
+	log.Debug().Str("method", "CreateTwin").Str("relay", relay).Str("pk", hex.EncodeToString(pk)).Msg("method called")
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.sub.CreateTwin(g.identity, relay, pk)
+}
+
+func (g *apiGateway) EnsureAccount(activationURL string, termsAndConditionsLink string, termsAndConditionsHash string) (info substrate.AccountInfo, err error) {
+	log.Debug().
+		Str("method", "EnsureAccount").
+		Str("activation url", activationURL).
+		Str("terms and conditions link", termsAndConditionsLink).
+		Str("terms and conditions hash", termsAndConditionsHash).
+		Msg("method called")
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.sub.EnsureAccount(g.identity, activationURL, termsAndConditionsLink, termsAndConditionsHash)
+}
+
+func (g *apiGateway) GetContract(id uint64) (result substrate.Contract, serr pkg.SubstrateError) {
+	log.Trace().Str("method", "GetContract").Uint64("id", id).Msg("method called")
+	contract, err := g.sub.GetContract(id)
+
+	serr = buildSubstrateError(err)
+	if err != nil {
+		return
+	}
+	return *contract, serr
+}
+
+func (g *apiGateway) GetContractIDByNameRegistration(name string) (result uint64, serr pkg.SubstrateError) {
+	log.Trace().Str("method", "GetContractIDByNameRegistration").Str("name", name).Msg("method called")
+	contractID, err := g.sub.GetContractIDByNameRegistration(name)
+
+	serr = buildSubstrateError(err)
+	return contractID, serr
+}
+
+func (g *apiGateway) GetFarm(id uint32) (substrate.Farm, error) {
+	log.Trace().Str("method", "GetFarm").Uint32("id", id).Msg("method called")
+	farm, err := g.sub.GetFarm(id)
+	if farm == nil {
+		farm = &substrate.Farm{}
+	}
+	return *farm, err
+}
+
+func (g *apiGateway) GetNode(id uint32) (substrate.Node, error) {
+	log.Trace().Str("method", "GetNode").Uint32("id", id).Msg("method called")
+	node, err := g.sub.GetNode(id)
+	if node == nil {
+		node = &substrate.Node{}
+	}
+	return *node, err
+}
```

Comment:
This code will work perfectly. I just think it's confusing to check the value for nil or not instead of the error. So again, while the result of the code is similar I just believe that the code below is less confusing to read and hard to break if it's needed to be modified later. What do you think?

```suggestion
func (g *apiGateway) GetNode(id uint32) (result substrate.Node, err error) {
	log.Trace().Str("method", "GetNode").Uint32("id", id).Msg("method called")
	node, err := g.sub.GetNode(id)
	if err != nil {
		return
	}

	return *node, err
}
```